utils/rjson.cc

```diff
@@ -236,6 +239,65 @@ std::string print(const rjson::value& value, size_t max_nested_level) {
     return std::string(buffer.GetString());
 }
 
+future<> print(const rjson::value& value, seastar::output_stream<char>& os, size_t max_nested_level) {
+    struct os_buffer {
+        seastar::output_stream<char>& _os;
+        temporary_buffer<char> _buf;
+        size_t _pos = 0;
+        future<> _f = make_ready_future<>();
+
+        using Ch = char;
+
+        void send(bool try_reuse = true) {
+            if (_f.failed()) {
+                _f.get0();
+            }
+            if (!_buf.empty() && _pos > 0) {
+                _buf.trim(_pos);
+                _pos = 0;
+                // Note: we're assuming we're writing to a buffered output_stream (hello http server).
+                // If we were not, or if (http) output_stream supported mixed buffered/packed content
+                // it might be a good idea to instead send our buffer as a packet directly. If so, the
+                // buffer size should probably increase (at least after first send()).
+                _f = _f.then([this, buf = std::move(_buf), &os = _os, try_reuse]() mutable -> future<> {
+                    return os.write(buf.get(), buf.size()).then([this, buf = std::move(buf), try_reuse]() mutable {
+                        // Chances are high we just copied this to output_stream buffer, and got here
+                        // immediately. If so, reuse the buffer.
+                        if (try_reuse && _buf.empty() && _pos == 0) {
+                            _buf = std::move(buf);
```

Comment:
I don't understand the benefit of this... There are two cases:
1. If _f is available and the write() is immediate, this will set _buf to what it already has, but the caller (Put) will simply overwrite it with a different buffer, so it's pointless.
2. If _f or write() are not immediately, this continuation will happen some time in the future. At that time _buf is already set to something else. So _buf = buf will free _buf and use buf instead. But what's the benefit of this over doing nothing - i.e., let _buf be and free buf?

So unless I'm missing something here, this "reuse" attempt (as it's written) is not doing what I think you hoped it would.

Reply:
Yeah, missing a code snippet below that I apparently merged away from my own work, we're supposed to only create new buffer if old one is empty. But the idea is to avoid allocs if we can.